File: Bachelor/Semester4/Systems_for_Design_and_Implementation/C#/ProiectCShMPP/Repository/repository/ReservationDbRepository.cs
using System;
using System.Data;
using ProiectCShMPP.domain;
using ProiectCShMPP.domain.validator;

namespace ProiectCShMPP.repository
{
    public class ReservationDbRepository : AbstractDBRepository<long, Reservation>, IReservationRepository
    {
        public ReservationDbRepository(IValidator<long, Reservation> validator) : base(validator)
        {
        }

        protected override IDbCommand GetSaveCommand(IDbConnection connection, Reservation entity)
        {
            var command = connection.CreateCommand();
            command.CommandText =
                "INSERT INTO reservations(client, phonenumber, tickets, tripid, userid) VALUES (@client, @phonenumber, @tickets, @tripdid, @userid);";

            var parameter = command.CreateParameter();
            parameter.ParameterName = "@client";
            parameter.Value = entity.Client;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@phonenumber";
            parameter.Value = entity.PhoneNumber;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@tickets";
            parameter.Value = entity.Tickets;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@tripdid";
            parameter.Value = entity.Trip.Id;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@userid";
            parameter.Value = entity.User.Id;
            command.Parameters.Add(parameter);
            
            return command;
        }

        protected override IDbCommand GetDeleteCommand(IDbConnection connection, long id)
        {
            var command = connection.CreateCommand();
            command.CommandText = "DELETE FROM reservations WHERE id=@id;";
            var parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            command.Parameters.Add(parameter);
            return command;
        }

        protected override IDbCommand GetUpdateCommand(IDbConnection connection, Reservation entity)
        {
            var command = connection.CreateCommand();
            command.CommandText =
                "UPDATE reservations SET client = @client, phonenumber = @phonenumber, tickets = @tickets, tripid = @tripid, userid = @userid WHERE id = @id;";

            var parameter = command.CreateParameter();
            parameter.ParameterName = "@client";
            parameter.Value = entity.Client;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@phonenumber";
            parameter.Value = entity.PhoneNumber;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@tickets";
            parameter.Value = entity.Tickets;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@tripid";
            parameter.Value = entity.Trip.Id;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@userid";
            parameter.Value = entity.User.Id;
            command.Parameters.Add(parameter);
            
            parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = entity.Id;
            command.Parameters.Add(parameter);

            return command;
        }

        protected override IDbCommand GetFindCommand(IDbConnection connection, long id)
        {
            var command = connection.CreateCommand();
            command.CommandText =
                "SELECT r.id, client, phonenumber, tickets, tripid, userid, username, password, destination, transportfirm," +
                " departuretime, price, seats FROM reservations r INNER JOIN users u on r.userid=u.id INNER JOIN trips t on r.tripid=t.id" +
                " WHERE r.id=@id;";
            var parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            command.Parameters.Add(parameter);
            return command;
        }

        protected override IDbCommand GetFindAllCommand(IDbConnection connection)
        {
            var command = connection.CreateCommand();
            command.CommandText =
                "SELECT r.id, client, phonenumber, tickets, tripid, userid, username, password, destination, transportfirm," +
                " departuretime, price, seats FROM reservations r INNER JOIN users u on r.userid=u.id INNER JOIN trips t on r.tripid=t.id;";
            return command;
        }

        protected override Reservation ExtractEntity(IDataReader reader)
        {
            _logger.InfoFormat("Entry Extract Entity");
            long id = (long) reader["id"];
            string client = (string) reader["client"];
            string phoneNumber = (string) reader["phonenumber"];
            int tickets = (int) reader["tickets"];

            long tripId = (long) reader["tripid"];
            Trip trip = new Trip(tripId)
            {
                Destination = (string) reader["destination"],
                TransportFirm = (string) reader["transportfirm"],
                DepartureTime = (DateTime) reader["departuretime"],
                Price = (float) (double) reader["price"],
                Seats = (int) reader["seats"]
            };

            long userId = (long) reader["userid"];
            User user = new User(userId)
            {
                Username = (string) reader["username"],
                Password = (string) reader["password"]
            };

            var reservation = new Reservation(id)
            {
                Client = client,
                PhoneNumber = phoneNumber,
                Tickets = tickets,
                Trip = trip,
                User = user
            };
            _logger.InfoFormat("Extracted {0}", reservation);
            return reservation;
        }
    }
}

File: Bachelor/Semester4/Systems_for_Design_and_Implementation/Java/Repository/src/main/java/repository/ReservationDBRepository.java
package repository;

import domain.Reservation;
import domain.Trip;
import domain.User;
import domain.validator.Validator;
import utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;

public class ReservationDBRepository extends AbstractDBRepository<Long, Reservation> implements ReservationRepository {

    public ReservationDBRepository(JdbcUtils jdbcUtils, Validator<Long, Reservation> validator) {
        super(jdbcUtils, validator);
    }

    @Override
    protected PreparedStatement getSaveStatement(Connection connection, Reservation entity) throws SQLException {
        PreparedStatement statement = connection.prepareStatement("INSERT INTO reservations(client, phonenumber, tickets, tripid, userid) VALUES (?,?,?,?,?);");
        statement.setString(1, entity.getClient());
        statement.setString(2, entity.getPhoneNumber());
        statement.setInt(3, entity.getTickets());
        statement.setLong(4, entity.getTrip().getId());
        statement.setLong(5, entity.getUser().getId());
        return statement;
    }

    @Override
    protected PreparedStatement getDeleteStatement(Connection connection, Long aLong) throws SQLException {
        PreparedStatement statement = connection.prepareStatement("DELETE FROM reservations WHERE id=?;");
        statement.setLong(1, aLong);
        return statement;
    }

    @Override
    protected PreparedStatement getUpdateStatement(Connection connection, Reservation entity) throws SQLException {
        PreparedStatement statement = connection.prepareStatement("UPDATE reservations " +
                "SET client=?,phonenumber=?,tickets=?,tripid=?,userid=? " +
                "WHERE id=?;");
        statement.setString(1, entity.getClient());
        statement.setString(2, entity.getPhoneNumber());
        statement.setInt(3, entity.getTickets());
        statement.setLong(4, entity.getTrip().getId());
        statement.setLong(5, entity.getUser().getId());
        statement.setLong(6, entity.getId());
        return statement;
    }

    @Override
    protected PreparedStatement getFindStatement(Connection connection, Long aLong) throws SQLException {
        PreparedStatement statement = connection.prepareStatement("SELECT r.id,client,phonenumber,tickets,tripid,userid,username,password,destination,transportfirm,departuretime,price,seats " +
                "FROM reservations r INNER JOIN users u on r.userid = u.id INNER JOIN trips t on r.tripid = t.id " +
                "WHERE r.id=?;");
        statement.setLong(1, aLong);
        return statement;
    }

    @Override
    protected PreparedStatement getFindAllStatement(Connection connection) throws SQLException {
        return connection.prepareStatement("SELECT r.id,client,phonenumber,tickets,tripid,userid,username,password,destination,transportfirm,departuretime,price,seats " +
                "FROM reservations r INNER JOIN users u on r.userid = u.id INNER JOIN trips t on r.tripid = t.id;");
    }

    @Override
    protected Reservation extractEntity(ResultSet resultSet) {
        logger.traceEntry();
        Reservation reservation = null;
        try{
            Long id = resultSet.getLong("id");
            String client = resultSet.getString("client");
            String phoneNumber = resultSet.getString("phonenumber");
            int tickets = resultSet.getInt("tickets");

            Long tripId = resultSet.getLong("tripid");
            String destination = resultSet.getString("destination");
            String transportFirm = resultSet.getString("transportfirm");
            LocalDateTime departureTime = resultSet.getTimestamp("departuretime").toLocalDateTime();
            float price = resultSet.getFloat("price");
            int seats = resultSet.getInt("seats");
            Trip trip = new Trip(tripId,destination,transportFirm,departureTime,price,seats);

            Long userId = resultSet.getLong("userid");
            String username = resultSet.getString("username");
            String password = resultSet.getString("password");
            User user = new User(userId, username, password);

            reservation = new Reservation(id,client,phoneNumber,tickets,trip,user);
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error "+e);
        }
        logger.traceExit(reservation);
        return reservation;
    }
}
